postgresql, stripe, migrations fixing

fabriziobertoglio1987 opened the pull request:
Development done: 
1) Adding postgresql
2) Editing the following migrations:
AddUserAndPostToVotes
remove_post_from_votes
3) Adding stripe gem and configuring stripe

schas002 commented on Gemfile:
```diff
@@ -38,12 +38,16 @@ gem 'sdoc', '~> 0.4.0', group: :doc
 
+# Gems for the project Mentor 
+
+gem 'stripe'
```
We don't need to pay for anything.

ArtOfCode- replied:
What's `stripe` used for, @fabriziobertoglio1987?

fabriziobertoglio1987 replied:
@ArtOfCode- the idea would be building a community like StackOverflow, but where user can pay a monthly subscription and get mentor help to find the solution to their problems.. 

So stripes would be for paying a monthly subscription

fabriziobertoglio1987 replied:
@schas002 @ArtOfCode-  I want to give you more info about this. There will be 2 type of users. Mentors will be able to earn money by helping other users solve their coding issues, Students will be paying a 20-50$ subscription to get help.

Payment can be also done in Bitcoin. 

I know that you don't like this aspect, but some developers may not be able to provide help for free, because they need to have a wage, on the otherside many students are learning ROR, but are challenged and need some directions..

I read your license, I hope this does not disturb you too much.. The project is open source. Other developers will be able to fork this, contribute and deploy their own instance.

Thanks 
Fabrizio

ArtOfCode- replied:
I'm fine with this, @fabriziobertoglio1987. Monetization ain't a crime.

schas002 commented on Gemfile:
```diff
@@ -1,10 +1,10 @@
 # Use sqlite3 as the database for Active Record
-gem 'sqlite3'
+# gem 'sqlite3'
+gem 'pg', '~> 0.20.0'
```
Suggestion: allow us to switch between SQLite and Postgres, since not all of us have a Postgre server at hand.

ArtOfCode- replied:
I'm gonna agree with this one. Unless there's some dependency reason we *have* to use PGSQL now, I'd rather either stick with default SQLite or use MySQL.

fabriziobertoglio1987 replied:
@ArtOfCode-  @schas002  I am wondering what is the best way to let the user choose the db ... do you mean just keep the database yml commented and instruction on how to implement it?

ArtOfCode- replied:
@fabriziobertoglio1987 Best way is probably as you suggested. Have both configurations in the database.yml file, with one commented out. Also include gems for both, so *both* `gem 'pg', '~> 0.20.0'` and `gem 'sqlite3'` in the Gemfile.

schas002 commented on Gemfile:
```diff
@@ -38,12 +38,16 @@ gem 'sdoc', '~> 0.4.0', group: :doc
 gem 'will_paginate', '~> 3.1.0'
 gem 'rails-html-sanitizer'
-gem 'letsencrypt_plugin'
+#gem 'letsencrypt_plugin'
```
[Not at all](https://blog.codinghorror.com/lets-encrypt-everything/).

fabriziobertoglio1987 replied:
I commented the gem 'letsencrypt_plugin' because heroku already provides encryption for you heroku domain..I had some problems with lets encrypt with my heroku server

schas002 commented on app/assets/javascripts/charges.coffee:
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Empty file.

fabriziobertoglio1987 replied:
@schas002 sorry for this horrible pull request. It is my first pull request, once performed I did not know how to remove this empty file. I guess I should do a `git reset file`

schas002 replied:
```sh
$ git rm file
$ git commit -m ':fire: remove file'
```

fabriziobertoglio1987 replied:
file deleted from the repository https://github.com/fabriziobertoglio1987/railsmentor/commit/36c8a41b58caa421d02225ca4ef8d03344c32cdb

schas002 commented on app/assets/stylesheets/charges.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the charges controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Empty file.

fabriziobertoglio1987 replied:
file deleted from the repository https://github.com/fabriziobertoglio1987/railsmentor/commit/36c8a41b58caa421d02225ca4ef8d03344c32cdb

schas002 commented on config/database.yml:
```diff
@@ -1,25 +1,27 @@
   timeout: 5000
+  username: fabrizio
+  password: <%= ENV['DATABASE_PW'] %>
+  #socket: /var/run/mysqld/mysqld.sock
```
Hardcoding `fabrizio` and `$DATABASE_PW` is not useful at all. As I said before, allow us to switch between SQLite and Postgres, and let us choose our own databases and credentials.

fabriziobertoglio1987 replied:
@schas002 ok... that is a good idea.. so maybe I keep a `database.yml` where both the sqlite and postgresql parameters are available...but they are just commented out...

schas002 commented on cutout:
```diff
@@ -0,0 +1 @@
+cutout
```
Unuseful file. Did you accidentally push this one?

ArtOfCode- replied:
$ rm cutout

fabriziobertoglio1987 replied:
@schas002 @ArtOfCode-  yes I did do many accidental things.. this should not be pushed to github.. I ll clean this up now

schas002 commented on test/controllers/charges_controller_test.rb:
```diff
@@ -0,0 +1,7 @@
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Empty file. An untested controller will probably bork as soon as you check it.

fabriziobertoglio1987 replied:
ok.. I need to test it then.. I'm not amazing at testing but really need to improve my Rspec skills

schas002 replied:
:blue_book: Stripe testing docs: https://stripe.com/docs/testing

schas002 commented:
Also, it would probably be better if your PR is from a feature branch. After the pull request, your fork's `master` branch would be borked, and you would need to recreate your fork all over.

ArtOfCode- commented on app/controllers/charges_controller.rb:
```diff
@@ -0,0 +1,25 @@
+class ChargesController < ApplicationController
+	def new
```
You've got an indentation error here; everything inside the class should start at 2 spaces indented.

fabriziobertoglio1987 replied:
file deleted from the repository  https://github.com/fabriziobertoglio1987/railsmentor/commit/36c8a41b58caa421d02225ca4ef8d03344c32cdb

ArtOfCode- commented on app/views/charges/create.html.erb:
```diff
@@ -0,0 +1 @@
+<h2>Thanks, you paid <strong>$5.00</strong>!</h2>
```
The $5.00 amount should probably be dynamic (as in, `<strong><%= @amount %></strong>` rather than hardcoded.

fabriziobertoglio1987 replied:
fixed https://github.com/fabriziobertoglio1987/railsmentor/commit/36c8a41b58caa421d02225ca4ef8d03344c32cdb

ArtOfCode- commented on app/views/charges/new.html.erb:
```diff
@@ -0,0 +1,18 @@
+      </div>
+    <% end %>
+    <label class="amount">
+      <span>Amount: $5.00</span>
```
Same here: this could be dynamic rather than hardcoded.

ArtOfCode- commented on db/migrate/20160430092324_add_default_false_value_to_notifications.rb:
```diff
@@ -1,5 +1,8 @@
 class AddDefaultFalseValueToNotifications < ActiveRecord::Migration
```
This doesn't need to be an up/down migration; the `change` method using a `change_column` call is able to handle forwards and backwards migrations.

fabriziobertoglio1987 replied:
@ArtOfCode- Y are right. no problem when I run `rake db:schema:load` so I restored the migrations as you had them https://github.com/fabriziobertoglio1987/railsmentor/commit/95df42dada12853633e5dece65b6787f3d725344

ArtOfCode- commented on db/migrate/20160501203952_add_default_to_is_deleted.rb:
```diff
@@ -1,6 +1,10 @@
 class AddDefaultToIsDeleted < ActiveRecord::Migration
-  def change
+  def up
```
Likewise, this can stay as `change`

ArtOfCode- commented on db/migrate/20160615175537_add_defaults_to_deleted.rb:
```diff
@@ -1,7 +1,10 @@
 class AddDefaultsToDeleted < ActiveRecord::Migration
-  def change
```
Same again

ArtOfCode- commented:
@schas002 Removing letsencrypt is probably a good idea, actually. Having TLS management mixed up with your application server is generally a separation-of-responsibility bad idea; HTTPS can be configured from the reverse proxy instead.

schas002 commented:
> _<...>_ it's difficult to test without issuing actual charges.

Stripe delivers: https://stripe.com/docs/testing

fabriziobertoglio1987 commented:
Thanks a lot Andrew. Tomorrow I review this all and I do all the fix you
require me to do. Thanks a lot for your experienced advise

Best Regards
Fabrizio

2017-08-18 17:05 GMT+07:00 Andrew Zyabin <notifications@github.com>:

> *<...>* it's difficult to test without issuing actual charges.
>
> Stripe delivers: https://stripe.com/docs/testing
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArtOfCode-/qpixel/pull/29#issuecomment-323314376>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AX1bF0SnVcluyvzNLBUI3od6r2D4YgRiks5sZWHbgaJpZM4O7IuR>
> .
>

fabriziobertoglio1987 commented:
@ArtOfCode-  actually what I am working on right now, I want to write some test with Rspec for the whole project, because I need to understand how the project work and maybe then try to upgrade to Rails 5. I also will do all the changes suggested

schas002 commented on public/assets/.sprockets-manifest-d0444265d8bed6e98efa1a64f58d5fa1.json:
```diff
@@ -0,0 +1 @@
+{"files":{"down-clear-137f65240ec6ef02feaf98d86122fb5308b8977161a37670ab6ef868d74de37e.png":{"logical_path":"down-clear.png","mtime":"2017-08-15T20:16:27+07:00","size":1279,"digest":"137f65240ec6ef02feaf98d86122fb5308b8977161a37670ab6ef868d74de37e","integrity":"sha256-E39lJA7G7wL+r5jYYSL7Uwi4l3Fho3Zwq274aNdN434="},"down-fill-49126c68b98ed5951765cf7d1ad82779fbe577e58e26aec9ca7152f792b5abce.png":{" ...
```
Vendored/minified file -- it might (might!) be helpful in prod, but it's never helpful in a development repository.

ArtOfCode- replied:
Meh, it doesn't exactly cause any harm.

schas002 commented on public/assets/application-0919e3b4bb6f47e19f6fe43bf31dc8f43fb841ba2de0e396ac9180b717533b6e.js:
```diff
@@ -0,0 +1,27 @@
+ * http://jquery.org/license
+ *
+ * Date: 2016-02-22T19:07Z
+ */
```
Minified file.

schas002 commented on public/assets/application-358bc2202961a3d3cf7b32db96a61f332c69dffcc530888f8fe916809e14d9fa.css:
```diff
@@ -0,0 +1 @@
+.question{width:90%;padding:10px;border-top:1px solid #bbb}.q-info-box{float:left;padding:0 10px}.vote-box{max-width:10%}.title-box{max-width:85%}p.large-text{font-size:1.4em}p.center{text-align:center}.vote-box{float:left;width:48px}.body-box{float:right;width:calc(100% - 63px)}.post-info{width:100%}.post-tags{width:50%;float:left}.post-author{width:45%;float:right;text-align:right}.users-row{wi ...
```
Minified file.

schas002 commented on public/assets/application-e19e7afeb43915e2165e0b0d4f927b842f594b39ea449a76251b4dd068dd7e79.js:
```diff
@@ -0,0 +1,12727 @@
+ * http://jquery.org/license
+ *
+ * Date: 2016-02-22T19:07Z
+ */
```
Bundling everything is never ever useful, at least in a development repository.

schas002 commented on public/assets/application-fbf25f87b816945058df04f4e29fe6317f0e0da6028c95c47944560f74bdf5a2.css:
```diff
@@ -0,0 +1,299 @@
+/* line 1, /home/fabrizio/Documents/Sublime/Rails/railsmentor/app/assets/stylesheets/questions.scss */
```
Bundled file. Also, personal information - since no one, except you, ever keeps files in `/home/fabrizio/Documents/Sublime/Rails/railsmentor`.

schas002 replied:
Also, making QPixel a paid mentor help app kinda breaks the spirit of QPixel IMO. It's God's gift for QPixel to be a FOSS clone of Stack Exchange, which is free for both the server and the client.

schas002 commented:
> You can run `rm public/assets/* -r` and another git commit to fix that; you could also add `public/assets` to `.gitignore`.

Do both.

fabriziobertoglio1987 commented:
Hello. I just have done `rm public/assets/* -r` then `git rm public/assets/* -r`
I added the folder to `.gitignore` then commited and pushed to my github repository.

Now the folder is clean

I am writing test cases with Rspec to  better understand the project, then I will try to migrate to `rails 5`, see if everything works.

Later I will try to configure correctly `stripe` and write test code.

Thanks a lot for your expert advise.
I will also remove from the repo other empty files.

This is my first fork.
Bye 
Fabrizio

schas002 commented:
> This is my first fork.

Eh, should have said earlier. You can learn how to do your first PR from this *free* series [How to Contribute to an Open Source Project on GitHub](https://egghead.io/series/how-to-contribute-to-an-open-source-project-on-github) (disclaimer: it's not mine)

ArtOfCode- commented:
Thanks, Fabrizio! I appreciate the effort you're putting into this. I'm on mobile right now, but I'll review the rest of your changes when I'm back at a desktop later on.

fabriziobertoglio1987 commented:
@ArtOfCode-  I did the modifications you requested, but now I found out that postgresql will have an error in your test suite.

> 1) Error:
UsersControllerTest#test_should_soft-delete_user:
SystemStackError: stack level too deep
    app/controllers/users_controller.rb:69:in `soft_delete'
    test/controllers/users_controller_test.rb:71:in `block in <class:UsersControllerTest>'


  2) Error:
UsersControllerTest#test_should_destroy_user:
SystemStackError: stack level too deep
    app/controllers/users_controller.rb:31:in `destroy'
    test/controllers/users_controller_test.rb:49:in `block in <class:UsersControllerTest>'

fabriziobertoglio1987 commented:
@ArtOfCode-  

I fixed the problem related to deleting the user. I run your test suite and now everything is fine. 

I created a association in `user.rb` of `has_many answers, :through => :questions`, then I added 
```
  has_many :comments, :dependent => :destroy
  has_many :flags, :dependent => :destroy
```
This fixed your test suite and everything looks ok now.

fabriziobertoglio1987 commented:
Hear I may change and use nested resources. For example `get questions/:question_id/ads/new`

and probably use the standard resource routing for the other routes

http://guides.rubyonrails.org/routing.html#nested-resources
```
  get    'questions/:id/answer',        :to => 'answers#new'
  post   'questions/:id/answer',        :to => 'answers#create'
  get    'answers/:id/edit',            :to => 'answers#edit'
  get    'questions/:id/answer',        :to => 'answers#new'
  post   'questions/:id/answer',        :to => 'answers#create'
  get    'answers/:id/edit',            :to => 'answers#edit'
  post   'answers/:id/edit',            :to => 'answers#update'
  patch  'answers/:id/edit',            :to => 'answers#update'
  delete 'answers/:id/delete',          :to => 'answers#destroy'
  patch  'answers/:id/delete',          :to => 'answers#undelete'
  post   'answers/:id/edit',            :to => 'answers#update'
  patch  'answers/:id/edit',            :to => 'answers#update'
  delete 'answers/:id/delete',          :to => 'answers#destroy'
  patch  'answers/:id/delete',          :to => 'answers#undelete'
```

ArtOfCode- commented:
@fabriziobertoglio1987 To your last comment: Redcarpet takes care of sanitisation by stepping everything but a subset of allowed tags. 

I like what you've done here. Again, I'll review when I'm on a desktop, but I think this is basically done.

fabriziobertoglio1987 commented:
@ArtOfCode-  Great. Thanks a lot!

In my last commit I added the resources routes for the questions and the nested resources for the answers in `routes.rb` , then I fixed the anwers controller and the answers controller test.

I'll keep writing test with rspec and then I will try to upgrade to Rails 5.

Thanks a lot
Fabrizio

fabriziobertoglio1987 commented:
Hi. I just found out my mistake.

`user` can not `has_many answers, :through => :questions` because this way we are saying that the question owner is the same from the answer owner. So I need to fix this and make sure it works on postgresql

Also this may have some impact on my and your test cases that I just fixed.
sorry 
Thanks 
Fabrizio

fabriziobertoglio1987 commented:
@ArtOfCode- 

I have another issue

`user has_many privileges` and `privilege has_and_belongs_to_many users`

This is causing problems with my `factories` test cases on `FactoryGirl` so  I want to remove this and change it in a `user has_and_belongs_to_many privileges` 

The fact is that the table `privileges_users` needs to have `belongs_to user` and `belongs_to privilege`, so I want to change your migration 

```
class CreateJoinTablePrivilegeUser < ActiveRecord::Migration
  def change
    create_join_table :privileges, :users do |t|
      t.index [:privilege_id, :user_id]
      t.index [:user_id, :privilege_id]
    end
  end
end
```
So I migrate down. But everything crashed. I mean, I **would to be able to migrate down the db**. I understand that it works if I do `rake db:schema:load` and thanks for explaining this to me, but we need to be able to migrate to a specific status of the db 

Anyway, I think the best is not to edit that migration, but just change it .. still migrating down to a specific point of the db may be useful for many collaborators

fabriziobertoglio1987 commented:
@ArtOfCode-  why should we not have the db migrations run smootly?

fabriziobertoglio1987 commented:
Hello @ArtOfCode- 

I believe we should be able to migrate the `db` at any state. Other Developer may want to fork this project and just do whether they want with it.

So I restored the original migrations I committed (that you asked me to change)...

fabriziobertoglio1987 commented:
for testing stripe
https://github.com/sb8244/rspec-stripe

ArtOfCode- commented on app/views/questions/show.html.erb:
```diff
@@ -15,5 +15,7 @@
 
 <% unless @question.is_closed %>
-  <a href="/questions/<%= @question.id %>/answer" class="btn btn-primary">Answer this Question</a>
+	<%= link_to new_question_answer_path(@question), class: "btn btn-primary" do %>Answer this Question<% end %>
```
`<%= link_to "Answer this Question", new_question_answer_path(@question), class: "btn btn-primary"  %>` also works

fabriziobertoglio1987 commented:
you should have written partials for the views and you should not include html code in `application.html.erb`

schas002 commented:
[Detecting what](https://github.com/ArtOfCode-/qpixel/pull/29/files#diff-6854d4e3d507b3d7cc33eef4e501c5c1)?

ArtOfCode- commented on app/assets/javascripts/user.js:
```diff
@@ -33,8 +31,8 @@ function hoverInEffect (div) {
-  h2 = $(div).children();
-  h2.removeClass('prices-h2-effect');
+  $number = $(div).find('.number')
+  $number.removeClass('prices-h2-effect');
```
Missing some `var` declarations here.

fabriziobertoglio1987 commented:
@ArtOfCode-  what do you mean with missing some `var` declarations?

ArtOfCode- commented:
You're declaring your JS variables as global variables at the moment, which pollutes the window scope. You should declare them as local variables with the var keyword.

schas002 commented:
@ArtOfCode- they aren't local, since they don't know closures. To declare closure aware local variables: `const` if your object doesn't change, `let` otherwise.

ArtOfCode- commented:
@schas002 Yes, I know ES6 exists. This project is ES5-only.

fabriziobertoglio1987 commented:
I clean up my projects from unuseful files in my `public/assets` once an year, but the code I keep it written as specified on the **rails best practices**, because in one year will be pretty hard to fix all the errors or remember all the improvement you should do, while cleaning the `public/assets` folder or deleting empty files from your git repository it is pretty repetitive easy task. No point of doing it now that I am always committing changes

I lost some times to figure out that you are really setting up badly your image. Just use the rails tags, what is this `<img src=''>` 

```
<img class="vote-button" data-vote="-1/<%= answer.id %>/a/<%= my_vote(answer).id %>"
             id="answer-<%= answer.id %>-down" src="/assets/down-fill.png" height="48" />
```

fabriziobertoglio1987 commented:
Right now I do not have the time to clean up my `github` repository, still I want to keep track hear of all the good changes and bugs that I fixed in your project. 

I changed your `<a src=''>` tags for the up and down icons in the questions and answers _expanded partial to be compatible with the asset pipeline

```
<%= image_tag 'up-clear.png', :class => 'vote-button', :id => "question-#{question.id}-up", :height => '48', :data => { :vote => "1/#{question.id}/q/-1" } %>      
        <!--img class="vote-button" data-vote="1/<%= question.id %>/q/-1"
           id="question-<%= question.id %>-up" src="/assets/up-clear.png" height="48" /-->
```

then I fixed the errors we had in your `application.js` file connected with turbolinks

this is included in my last commit

using var in that function stops the function from working so, I ll try to fix that later, but I don't have it in my priority things to do right now

ArtOfCode- commented:
You can probably add `public/assets` to `.gitignore`, @fabriziobertoglio1987 - there's no real need to have them in the repository.

fabriziobertoglio1987 commented:
@ArtOfCode-  yes your right. soon the app will be finalized and I can clean up all the folders.

schas002 commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2016 QPixel Contributors
+Copyright (c) 2017 Fabrizio Bertoglio
```
**Unsolicitedly moving copyright to you exclusively is illegal**. The 'QPixel Contributors' would have to move the copyright to you themselves, and since that's multiple persons, well...

fabriziobertoglio1987 commented:
@schas002 that was a mistake. I did not want to include my name in the license, but github required the name. I just included `QPixel Contributors`. It needs to have a license.
